### DISCUSSION TITLE
Alpine Docker container

### DISCUSSION THREAD

**Discussion by coolapso** [pivpn/pivpn#1584] [Q&A]
checking my options to implement CI testing for alpine and during the discussions to add Alpine support, I Was under impression that some testing on docker containers was also done. 
considering it was the main intent of supporting alpine goes exactly towards the objective of getting pivpn somehow working on docker containers with alpine, so I decided to go on some naive testing and just start an `alpine:latest` container and run pivpn script on it. most of the script worked without errors except this, and obviously wireguard service does not start. 

```
Executing busybox-1.35.0-r15.trigger
OK: 275 MiB in 141 packages
:::    Package wireguard-tools successfully installed!
:::    Package libqrencode successfully installed!
::: Using Quad9 servers.
::: Adding wg-quick unit
::: Backing up the wireguard folder to /etc/wireguard_2022-08-06-181411.tar.gz
::: Server Keys have been generated.
::: Server config generated.
sysctl: error setting key 'net.ipv4.ip_forward': Read-only file system
iptables v1.8.8 (legacy): can't initialize iptables table `nat': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
getsockopt failed strangely: Operation not permitted
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
getsockopt failed strangely: Operation not permitted
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
getsockopt failed strangely: Operation not permitted
getsockopt failed strangely: Operation not permitted
::: Install Complete...
::: Restarting services...
 * WARNING: wg-quick is already starting
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/community/x86_64/APKINDEX.tar.gz
(1/7) Installing libffi (3.4.2-r1)
(2/7) Installing libucontext (1.2-r0)
(3/7) Installing yaml (0.2.5-r0)
(4/7) Installing ruby-libs (3.1.2-r0)
```

This errors are totally expected as the networking inside the containers is not exactly the same as with a native OS, so this are the errors i encountered earlier in my first attempts to get pivpn running on docker and which I got through to some extent but did not had time to finish. 

@giulio-coa Did I understood wrong about docker (is a possibility since all of it was quite busy time for me)? isn't it working yet on docker or did you get it to run on docker and there is something wrong?  

"baremetal" Alpine is working just fine, however with some caveats that I will add later to the documentation. 

AWS AMI's privilege escalation needs to be done with `doas`  otherwise its impossible to install dependencies. 

```
doas su
apk add bash
apk add busybox-initscripts
apk add sudo
echo '%wheel ALL=(ALL) ALL' > /etc/sudoers.d/wheel
mv wheel /etc/sudoers.d/
chown root:root /etc/sudoers.d/wheel
```

**Comment by ghost** [comment]
Did you use option `--cap-add NET_ADMIN` ?

**Reply by coolapso** [reply]
As i said .. i just went on a "naive approach" and the goal was exactly to start discussion and find out what needs to be documented :) let me try and will get back.

**Reply by coolapso** [reply]
ok .. that part seems fine and working .. next step. 

```
/ # pivpn add -n test2
::: Client Keys generated
::: Client config generated
::: Updated server config
 * You are attempting to run an openrc service on a
 * system which openrc did not boot.
 * You may be inside a chroot or you may have used
 * another initialization system to boot this system.
 * In this situation, you will get unpredictable results!
 * If you really want to do this, issue the following command:
 * touch /run/openrc/softlevel
 * ERROR: wg-quick failed to start
::: Failed to reload WireGuard
======================================================================
::: Done! test2.conf successfully created!
::: test2.conf was copied to /home/test/configs for easy transfer.
::: Please use this profile only on one device and create additional
::: profiles for other devices. You can also use pivpn -qr
::: to generate a QR Code you can scan with the mobile app.
======================================================================
```

This is not by all means how it should run on a docker image, however just trying in the most basic way and trying to pick up the pieces and figuring out what is working and not.

**Reply by ghost** [reply]
https://github.com/gliderlabs/docker-alpine/issues/437#issuecomment-494200575

**Reply by ghost** [reply]
I was waiting to close PR #1572 and then offer you an official repo for docker, but you are anticipating me hahahaha

**Reply by coolapso** [reply]
well, i had some time now, so i just decided to pave some ground and make some testing to get the CI Pipelines after we merge that one. Still getting the same error after mounting the volume tho.

**Reply by coolapso** [reply]
as for an official docker repo, there is still a long way to go .. but initially I think we can provide a way for users to build their images. providing built images is still to dangerous, as all the encryption bits and pieces need to be done separately do avoid ending up with multiple users with the same ciphers.

**Reply by ghost** [reply]
> as for an official docker repo, there is still a long way to go .. but initially I think we can provide a way for users to build their images. providing built images is still to dangerous, as all the encryption bits and pieces need to be done separately do avoid ending up with multiple users with the same ciphers. 

I thought I was doing something like [this](https://github.com/InnovativeInventor/docker-pivpn/blob/master/amd64/Dockerfile)

**Reply by ghost** [reply]
> well, i had some time now, so i just decided to pave some ground and make some testing to get the CI Pipelines after we merge that one. Still getting the same error after mounting the volume tho.  

You must add these row to your Dockerfile
```
VOLUME [ "/sys/fs/cgroup" ]

RUN mkdir -p /run/openrc \
  && touch /run/openrc/softlevel
```

**Reply by coolapso** [reply]
> > as for an official docker repo, there is still a long way to go .. but initially I think we can provide a way for users to build their images. providing built images is still to dangerous, as all the encryption bits and pieces need to be done separately do avoid ending up with multiple users with the same ciphers.
> 
> I thought I was doing something like [this](https://github.com/InnovativeInventor/docker-pivpn/blob/master/amd64/Dockerfile)

for someone to use and build their own image its okay. Its not okay however to provide the images publicly based  on that.

**Reply by ghost** [reply]
> for someone to use and build their own image its okay. Its not okay however to provide the images publicly based on that.

Maybe we can do two different branches
Something like `docker-pivpn/master` for the complete release and `docker-pivpn/base-build` for the release to build

**Reply by coolapso** [reply]
doesn't feel that manageable. 

When i started testing a possible pivpn docker image, what came to my mind was to have a `DOCKER=1` Variable when building docker images which would tell pivpn script to only install the components without generating any of the certificates/cryptography components and those would be generated on runtime when starting a container.

IF a volume/bind mount exists and contains the stuff it needs then it skips it, otherwise generates them. 

but this was just a concept never materialized it.

**Reply by coolapso** [reply]
btw this worked: 

```
VOLUME [ "/sys/fs/cgroup" ]

RUN mkdir -p /run/openrc \
  && touch /run/openrc/softlevel
``` 

So I think ti can be used to make the CI testing, as neither Github or Travis CI provide any Alpine image to use with and I don't really wanna go through the steps of getting a runner just for that, it would be something extra to maintain and I have already plenty :p

**Reply by ghost** [reply]
> doesn't feel that manageable.
> 
> When i started testing a possible pivpn docker image, what came to my mind was to have a `DOCKER=1` Variable when building docker images which would tell pivpn script to only install the components without generating any of the certificates/cryptography components and those would be generated on runtime when starting a container.
> 
> IF a volume/bind mount exists and contains the stuff it needs then it skips it, otherwise generates them.
> 
> but this was just a concept never materialized it.

We could do a multi-stage container
If the user uses option `--build-arg COMPLETE=1`, then create the full version, otherwise the one to build

**Reply by coolapso** [reply]
That can also be an idea. but i'd rather find a way to keep users hands out of it :p 
let's wait till @orazioedoardo makes his review of https://github.com/pivpn/pivpn/pull/1572, otherwise ill merge it during the week.
then Ill make the CI changes plus some other changes i have planed, and then after maybe we can finally come with an official pivpn image.

**Reply by ghost** [reply]
Ok

**Reply by ghost** [reply]
> So I think it can be used to make the CI testing, as neither Github or Travis CI provide any Alpine image to use with and I don't really wanna go through the steps of getting a runner just for that, it would be something extra to maintain and I have already plenty :p

We could do our test
Pi-hole use `tox` to do this
To add a new distro, you must only add its name to the [distro matrix](https://github.com/pi-hole/pi-hole/blob/development/.github/workflows/test.yml#L46), create the opportune new files (`Dockerfile`, tox `.ini` file) into the [test directory](https://github.com/pi-hole/pi-hole/tree/development/test) and adapt the existing tests if necessary
I have tried this procedure in order to add Alpine and it works
It's simple and often you don't have to edit old files
